include/mfmg/amge.templates.hpp

```diff
@@ -0,0 +1,200 @@
+/*************************************************************************
+ * Copyright (c) 2017 by the mfmg authors                                *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * This file is part of the mfmg libary. mfmg is distributed under a BSD *
+ * 3-clause license. For the licensing terms see the LICENSE file in the *
+ * top-level directory                                                   *
+ *                                                                       *
+ * SPDX-License-Identifier: BSD-3-Clause                                 *
+ *************************************************************************/
+
+#include <mfmg/amge.hpp>
+
+#include <deal.II/base/work_stream.h>
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/numerics/data_out.h>
+
+#include <fstream>
+
+namespace mfmg
+{
+template <int dim, typename VectorType>
+AMGe<dim, VectorType>::AMGe(MPI_Comm comm, dealii::DoFHandler<dim> &dof_handler)
+    : _comm(comm), _dof_handler(dof_handler)
+{
+}
+
+template <int dim, typename VectorType>
+unsigned int AMGe<dim, VectorType>::build_agglomerates(
+    std::array<unsigned int, dim> const &agglomerate_dim)
+{
+  // Faces in deal.II are orderd as follows: left (x_m) = 0, right (x_p) = 1,
+  // front (y_m) = 2, back (y_p) = 3, bottom (z_m) = 4, top (z_p) = 5
+  unsigned int constexpr x_p = 1;
+  unsigned int constexpr y_p = 3;
+  unsigned int constexpr z_p = 5;
+
+  // Flag the cells to create the agglomerates
+  unsigned int agglomerate = 1;
+  for (auto cell : _dof_handler.active_cell_iterators())
+  {
+    if ((cell->is_locally_owned()) && (cell->user_index() == 0))
+    {
+      cell->set_user_index(agglomerate);
+      auto current_z_cell = cell;
+      unsigned int const d_3 = (dim < 3) ? 1 : agglomerate_dim.back();
+      for (unsigned int i = 0; i < d_3; ++i)
+      {
+        auto current_y_cell = current_z_cell;
+        for (unsigned int j = 0; j < agglomerate_dim[1]; ++j)
+        {
+          auto current_cell = current_y_cell;
+          for (unsigned int k = 0; k < agglomerate_dim[0]; ++k)
+          {
+            // TODO For now, we assume that there is no adaptive refinement
+            current_cell->set_user_index(agglomerate);
+            if (current_cell->at_boundary(x_p) == false)
+            {
+              auto neighbor_cell = current_cell->neighbor(x_p);
+              if ((neighbor_cell->is_locally_owned()) &&
+                  (neighbor_cell->user_index() == 0))
```

Comment:
Because I am trying to add a cell to the agglomerate but I cannot do it if it is already part of an other agglomerate. Otherwise, the first agglomerate built will lose cells to other agglomerate.

Reply:
But this should never happen on a uniform mesh. If you are in a situation where you get to this situation, you might as well throw as you no longer know what's going on, right? Your constructed aggregate is unlikely to be a brick any longer.

Reply:
Yes you are right.